src/m365/spo/commands/applicationcustomizer/applicationcustomizer-add.ts

```diff
@@ -0,0 +1,224 @@
+import * as spoListItemAddCommand from '../listitem/listitem-add';
+import * as spoListItemListCommand from '../listitem/listitem-list';
+import * as spoCustomActionAddCommand from '../customaction/customaction-add';
+import { urlUtil } from '../../../../utils/urlUtil';
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  title: string;
+  webUrl?: string;
+  clientSideComponentId: string;
+  clientSideComponentProperties?: string;
+  webTemplate?: string;
+}
+
+class SpoApplicationCustomizerAddCommand extends SpoCommand {
+  private location = 'ClientSideExtension.ApplicationCustomizer';
+  public get name(): string {
+    return commands.APPLICATIONCUSTOMIZER_ADD;
+  }
+
+  public get description(): string {
+    return 'Add an application customizer to a site';
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        clientSideComponentProperties: typeof args.options.clientSideComponentProperties !== 'undefined',
+        webTemplate: typeof args.options.webTemplate !== 'undefined',
+        webUrl: typeof args.options.webUrl !== 'undefined'
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '-t, --title <title>'
+      },
+      {
+        option: '-u, --webUrl [webUrl]'
+      },
+      {
+        option: '-i, --clientSideComponentId <clientSideComponentId>'
+      },
+      {
+        option: '--clientSideComponentProperties [clientSideComponentProperties]'
+      },
+      {
+        option: '--webTemplate [webTemplate]'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        if (!validation.isValidGuid(args.options.clientSideComponentId)) {
+          return `${args.options.clientSideComponentId} is not a valid GUID`;
+        }
+
+        if (args.options.webUrl) {
+          const isValidSharePointUrl: boolean | string = validation.isValidSharePointUrl(args.options.webUrl);
+          if (isValidSharePointUrl !== true) {
+            return isValidSharePointUrl;
+          }
+        }
+
+        if (args.options.webUrl && args.options.webTemplate) {
+          return `The options 'webUrl' and 'webTemplate' cannot be set together`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    try {
+      if (this.verbose) {
+        logger.logToStderr(`Adding application customizer with ${args.options.clientSideComponentId} to ${args.options.webUrl ? 'specific web' : 'tenant app catalog'}`);
+      }
+
+      if (args.options.webUrl) {
+        const options: spoCustomActionAddCommandOptions = {
```

Comment:
Great point. I'll rework this!

Reply:
@martinlingstuyl Would I have to check in both the tenant app catalog and, if no solution is found, the site collection app catalog, or what would you suggest?

Reply:
That's a very good point! We need to check the site col specific app catalog as well when using webUrl 🤦🏼‍♂️ totally forgot. Please do implement it!

Reply:
Hi @martinlingstuyl 
How would you go about this then? Currently, we will throw errors if a solution in not found in the tenant-app catalog. However, for the component to work in a webUrl, there will have to be a solution installed on the site or the clientsidecomponent will not be available. 
I've also noticed that there is a list 'ComponentManifest', but this list is only available when there is a site collection app catalog enabled on the site, so the components would get in there when the application is installed. I'm not certain how to correctly implement this really.

Reply:
Hmm, you're correct. Implementing this will lead us into a forest of http calls. Which will downgrade performance and not add a big enough benefit. I think we are making this too difficult. Let's retrace our steps and solve the issue at hand. 

- Let's not check the availability of the appcustomizer when deploying on a specific site. For now... we can always change it in the future if we have a good idea on how to do it.

- also: the reason we implemented the checks on the tenant wide path was because it would throw an error on adding an item to the tenant wide extension list if the solution was not available. To keep things consistent between the two code paths I think we should just remove the checks we're doing and just wrap the http call to the tenant wide list within a try catch statement. We can throw an error in the catch block about what the probably reason for the error is. Which is actually fine as well. As long as the user has some extra information. This will also be faster.

I'm sorry to have lead you down this path and you investing your time into it, only to discover now that it's better to take a different road 😅

Reply:
Do you agree with me by the way? We could ask some extra advice from a third person.

Reply:
I agree that building the checks for a site collection custom action would be really difficult. I'll reformat the code and will check what the error message is when no component is found so that we can display this to the user. Will met you know

Reply:
Let's sleep on it and ask someone else first. I'll draft something...